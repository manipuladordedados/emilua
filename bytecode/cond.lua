OUTPUT = arg[#arg]

function strip_xxd_hdr(file)
    local lines = {}
    for line in file:lines() do
        lines[#lines + 1] = line
    end
    table.remove(lines, 1)
    lines[#lines] = nil
    lines[#lines] = nil
    return table.concat(lines)
end

function cond_wait_bootstrap(error, native)
    return function(cnd, mtx)
        local e = native(cnd, mtx)
        mtx:lock()
        if e then
            error(e)
        end
    end
end

cond_wait_bytecode = string.dump(cond_wait_bootstrap, true)
f = io.open(OUTPUT, 'wb')
f:write(cond_wait_bytecode)
f:close()
cond_wait_cdef = strip_xxd_hdr(io.popen('xxd -i ' .. OUTPUT))

f = io.open(OUTPUT, 'wb')

f:write([[
#include <cstddef>

namespace emilua {
unsigned char cond_wait_bytecode[] = {
]])

f:write(cond_wait_cdef)

f:write('};')
f:write(string.format('std::size_t cond_wait_bytecode_size = %i;',
                      #cond_wait_bytecode))

f:write('} // namespace emilua')
