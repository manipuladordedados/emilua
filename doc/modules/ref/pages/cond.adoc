= cond

ifeval::["{doctype}" == "manpage"]

== Name

Emilua - Lua execution engine

== Synopsis

endif::[]

[source,lua]
----
local cond = require('cond')

local function queue_consumer()
    scope(function()
        scope_cleanup_push(function() queue_mtx:unlock() end)
        queue_mtx:lock()
        while #queue == 0 do
            queue_cond:wait(queue_mtx)
        end
        for _, e in ipairs(queue) do
            consume_item(e)
        end
        queue = {}
    end)
end
----

A condition variable.

== Functions

=== `new() -> cond`

Constructor.

=== `wait(self, m: mutex)`

Read `pthread_cond_wait()`

`wait()` is an interruption point. Prior to the delivery of the interruption
request, the underlying mutex is re-acquired under the hood.

=== `notify_all(self)`

Read `pthread_cond_broadcast()`.

=== `notify_one(self)`

Read `pthread_cond_signal()`.

== Notifying without a lock

If the condition variable, the notifier fiber and the waiting fiber all run in
the same thread (and cooperative multitasking is used instead preemptive
multitasking), then there is enough level of determinism to lift one restriction
that exists in traditional condition variables.

[quote,'<https://en.cppreference.com/w/cpp/thread/condition_variable>']
____
Even if the shared variable is atomic, it must be modified under the mutex in
order to correctly publish the modification to the waiting thread.
____

The reason why this restriction on the notifier fiber/thread exists is to avoid
a race. Consider the following waiter fiber and the notifier fiber:

[source,lua]
----
local function consumer()
    scope(function()
        scope_cleanup_push(function() m:unlock() end)
        m:lock()
        while not ready do
            c:wait(m)
        end

        -- ...
    end)
end

local function producer()
    ready = true
    c:notify_one()
end
----

Pay attention to the points when the waiter fiber checks if the event has been
signalled by testing `ready` and the instant it blocks on `c.wait()`. If the
notifier fiber mutates the shared variable and calls `c.notify_one()` between
these two points, then the signalization is lost. `c.notify_one()` would be
called by the time there would be no fiber blocked on `c.wait()`. Thatâ€™s why the
notifier fiber need to mutate the shared variable through a mutex.

In Emilua, this restriction doesn't apply (as long as there are no suspension
points between the time the waiting fiber tests the condition and calls
`c.wait()`) and the notifier fiber can mutate the shared variable without
holding a lock on the mutex. In this case, the condition variable essentially
becomes a non-suspending way (post semantics) to unpark a parked fiber (yes,
I've exploited this property in the past to avoid a few round-trips).
