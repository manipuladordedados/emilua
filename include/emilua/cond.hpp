#pragma once

#include <emilua/core.hpp>

namespace emilua {

extern char cond_key;

void init_cond_module(lua_State* L);

} // namespace emilua
