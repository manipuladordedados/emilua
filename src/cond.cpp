#include <emilua/dispatch_table.hpp>
#include <emilua/mutex.hpp>
#include <emilua/cond.hpp>

namespace emilua {

extern unsigned char cond_wait_bytecode[];
extern std::size_t cond_wait_bytecode_size;

char cond_key;
static char cond_mt_key;
static char cond_wait_key;

struct cond_handle
{
    std::deque<lua_State*> pending;
};

static int cond_wait(lua_State* L)
{
    auto cond_handle = reinterpret_cast<struct cond_handle*>(
        lua_touserdata(L, 1));
    if (!cond_handle || !lua_getmetatable(L, 1)) {
        push(L, std::errc::invalid_argument);
        lua_pushliteral(L, "arg");
        lua_pushinteger(L, 1);
        lua_rawset(L, -3);
        return lua_error(L);
    }
    rawgetp(L, LUA_REGISTRYINDEX, &cond_mt_key);
    if (!lua_rawequal(L, -1, -2)) {
        push(L, std::errc::invalid_argument);
        lua_pushliteral(L, "arg");
        lua_pushinteger(L, 1);
        lua_rawset(L, -3);
        return lua_error(L);
    }

    auto mutex_handle = reinterpret_cast<struct mutex_handle*>(
        lua_touserdata(L, 2));
    if (!mutex_handle || !lua_getmetatable(L, 2)) {
        push(L, std::errc::invalid_argument);
        lua_pushliteral(L, "arg");
        lua_pushinteger(L, 1);
        lua_rawset(L, -3);
        return lua_error(L);
    }
    rawgetp(L, LUA_REGISTRYINDEX, &mutex_mt_key);
    if (!lua_rawequal(L, -1, -2)) {
        push(L, std::errc::invalid_argument);
        lua_pushliteral(L, "arg");
        lua_pushinteger(L, 2);
        lua_rawset(L, -3);
        return lua_error(L);
    }

    if (!mutex_handle->locked) {
        push(L, std::errc::operation_not_permitted);
        return lua_error(L);
    }

    EMILUA_CHECK_SUSPEND_ALLOWED(mutex_handle->vm_ctx, L);

    lua_pushvalue(L, 1);
    lua_pushlightuserdata(L, mutex_handle->vm_ctx.current_fiber());
    lua_pushcclosure(
        L,
        [](lua_State* L) -> int {
            auto vm_ctx = get_vm_context(L).shared_from_this();
            auto handle = reinterpret_cast<struct cond_handle*>(
                lua_touserdata(L, lua_upvalueindex(1)));
            auto current_fiber = reinterpret_cast<lua_State*>(
                lua_touserdata(L, lua_upvalueindex(2)));

            auto it = std::find(handle->pending.begin(), handle->pending.end(),
                                current_fiber);
            if (it == handle->pending.end()) {
                // notify() was called before interrupt()
                //
                // an alternative would be to clear the interrupter on notify(),
                // but this approach works well enough
                return 0;
            }

            handle->pending.erase(it);
            vm_ctx->strand().post([vm_ctx,current_fiber]() {
                vm_ctx->fiber_prologue(current_fiber);
                push(current_fiber, errc::interrupted);
                vm_ctx->reclaim_reserved_zone();
                int res = lua_resume(current_fiber, 1);
                vm_ctx->fiber_epilogue(res);
            }, std::allocator<void>{});
            return 0;
        },
        2);
    set_interrupter(L);

    cond_handle->pending.emplace_back(mutex_handle->vm_ctx.current_fiber());

    // mutex:unlock() {{{
    if (mutex_handle->pending.size() == 0) {
        mutex_handle->locked = false;
    } else {
        auto vm_ctx = mutex_handle->vm_ctx.shared_from_this();
        auto next = mutex_handle->pending.front();
        mutex_handle->pending.pop_front();
        vm_ctx->strand().post([vm_ctx,next]() {
            vm_ctx->fiber_resume_trivial(next);
        }, std::allocator<void>{});
    }
    // }}}

    return lua_yield(L, 0);
}

static int cond_notify_one(lua_State* L)
{
    auto handle = reinterpret_cast<cond_handle*>(lua_touserdata(L, 1));
    if (!handle || !lua_getmetatable(L, 1)) {
        push(L, std::errc::invalid_argument);
        lua_pushliteral(L, "arg");
        lua_pushinteger(L, 1);
        lua_rawset(L, -3);
        return lua_error(L);
    }
    rawgetp(L, LUA_REGISTRYINDEX, &cond_mt_key);
    if (!lua_rawequal(L, -1, -2)) {
        push(L, std::errc::invalid_argument);
        lua_pushliteral(L, "arg");
        lua_pushinteger(L, 1);
        lua_rawset(L, -3);
        return lua_error(L);
    }

    if (handle->pending.size() == 0)
        return 0;

    auto vm_ctx = get_vm_context(L).shared_from_this();
    auto next = handle->pending.front();
    handle->pending.pop_front();
    vm_ctx->strand().post([vm_ctx,next]() {
        vm_ctx->fiber_resume(next);
    }, std::allocator<void>{});
    return 0;
}

static int cond_notify_all(lua_State* L)
{
    auto handle = reinterpret_cast<cond_handle*>(lua_touserdata(L, 1));
    if (!handle || !lua_getmetatable(L, 1)) {
        push(L, std::errc::invalid_argument);
        lua_pushliteral(L, "arg");
        lua_pushinteger(L, 1);
        lua_rawset(L, -3);
        return lua_error(L);
    }
    rawgetp(L, LUA_REGISTRYINDEX, &cond_mt_key);
    if (!lua_rawequal(L, -1, -2)) {
        push(L, std::errc::invalid_argument);
        lua_pushliteral(L, "arg");
        lua_pushinteger(L, 1);
        lua_rawset(L, -3);
        return lua_error(L);
    }

    auto vm_ctx = get_vm_context(L).shared_from_this();
    for (auto& p: handle->pending) {
        vm_ctx->strand().post([vm_ctx,p]() {
            vm_ctx->fiber_resume(p);
        }, std::allocator<void>{});
    }
    handle->pending.clear();
    return 0;
}

static int cond_mt_index(lua_State* L)
{
    return dispatch_table::dispatch(
        hana::make_tuple(
            hana::make_pair(
                BOOST_HANA_STRING("notify_all"),
                [](lua_State* L) -> int {
                    lua_pushcfunction(L, cond_notify_all);
                    return 1;
                }
            ),
            hana::make_pair(
                BOOST_HANA_STRING("wait"),
                [](lua_State* L) -> int {
                    rawgetp(L, LUA_REGISTRYINDEX, &cond_wait_key);
                    return 1;
                }
            ),
            hana::make_pair(
                BOOST_HANA_STRING("notify_one"),
                [](lua_State* L) -> int {
                    lua_pushcfunction(L, cond_notify_one);
                    return 1;
                }
            )
        ),
        [](std::string_view /*key*/, lua_State* L) -> int {
            push(L, errc::bad_index);
            lua_pushliteral(L, "index");
            lua_pushvalue(L, 2);
            lua_rawset(L, -3);
            return lua_error(L);
        },
        tostringview(L, 2),
        L
    );
}

static int cond_new(lua_State* L)
{
    auto buf = reinterpret_cast<cond_handle*>(
        lua_newuserdata(L, sizeof(cond_handle))
    );
    rawgetp(L, LUA_REGISTRYINDEX, &cond_mt_key);
    int res = lua_setmetatable(L, -2);
    assert(res); boost::ignore_unused(res);
    new (buf) cond_handle{};
    return 1;
}

void init_cond_module(lua_State* L)
{
    lua_pushlightuserdata(L, &cond_key);
    lua_newtable(L);
    {
        lua_createtable(L, /*narr=*/0, /*nrec=*/3);

        lua_pushliteral(L, "__metatable");
        lua_pushlightuserdata(L, nullptr);
        lua_rawset(L, -3);

        lua_pushliteral(L, "__index");
        lua_pushcfunction(
            L,
            [](lua_State* L) -> int {
                auto key = tostringview(L, 2);
                if (key == "new") {
                    lua_pushcfunction(L, cond_new);
                    return 1;
                } else {
                    push(L, errc::bad_index);
                    lua_pushliteral(L, "index");
                    lua_pushvalue(L, 2);
                    lua_rawset(L, -3);
                    return lua_error(L);
                }
            });
        lua_rawset(L, -3);

        lua_pushliteral(L, "__newindex");
        lua_pushcfunction(
            L,
            [](lua_State* L) -> int {
                push(L, std::errc::operation_not_permitted);
                return lua_error(L);
            });
        lua_rawset(L, -3);
    }
    lua_setmetatable(L, -2);
    lua_rawset(L, LUA_REGISTRYINDEX);

    lua_pushlightuserdata(L, &cond_mt_key);
    {
        lua_createtable(L, /*narr=*/0, /*nrec=*/3);

        lua_pushliteral(L, "__metatable");
        lua_pushliteral(L, "cond");
        lua_rawset(L, -3);

        lua_pushliteral(L, "__index");
        lua_pushcfunction(L, cond_mt_index);
        lua_rawset(L, -3);

        lua_pushliteral(L, "__gc");
        lua_pushcfunction(L, finalizer<cond_handle>);
        lua_rawset(L, -3);
    }
    lua_rawset(L, LUA_REGISTRYINDEX);

    {
        lua_pushlightuserdata(L, &cond_wait_key);
        int res = luaL_loadbuffer(
            L, reinterpret_cast<char*>(cond_wait_bytecode),
            cond_wait_bytecode_size, nullptr);
        assert(res == 0); boost::ignore_unused(res);
        lua_pushcfunction(L, lua_error);
        lua_pushcfunction(L, cond_wait);
        lua_call(L, 2, 1);
        lua_rawset(L, LUA_REGISTRYINDEX);
    }
}

} // namespace emilua
