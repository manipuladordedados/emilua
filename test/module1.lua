-- only_main_fiber_may_import

spawn(function()
    require('cond')
end)
