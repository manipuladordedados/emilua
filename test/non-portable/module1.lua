-- A future Emilua version might allow module renaming for global modules

require('../cond')
