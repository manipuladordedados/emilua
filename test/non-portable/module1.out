Main fiber from VM 0x1 panicked: 'Module not found'
stack traceback:
	[string "?"]: in function 'require'
	input.lua:3: in main chunk
	[C]: in function ''
	[string "?"]: in function <[string "?"]:0>
